provider/oracle/provider.go

```diff
@@ -0,0 +1,175 @@
+package oracle
+
+import (
+	"github.com/juju/errors"
+	"github.com/juju/jsonschema"
+	"github.com/juju/loggo"
+
+	oci "github.com/hoenirvili/go-oracle-cloud/api"
+	"github.com/juju/juju/cloud"
+	"github.com/juju/juju/environs"
+	"github.com/juju/juju/environs/config"
+)
+
+var logger = loggo.GetLogger("juju.provider.oracle")
+
+const (
+	providerType = "oracle"
+)
+
+// provide friendly aliases for the register provider function
+var providerAliases = []string{"ocl", "orcl", "oci"}
+
+// environProvider type implements environs.EnvironProvider interface
+// this will represent a computing and storage provider of the orcale cloud
+// alongiside environs.EnvironProvider this implements config.Validator interface and
+// environs.ProviderCredentials also.
+type environProvider struct {
+	// client for oracle cloud api connection
+	client *oci.Client
+}
+
+// CloudSchema returns the schema used to validate input for add-cloud.  Since
+// this provider does support custom clouds, this always returns non-nil
+func (e environProvider) CloudSchema() *jsonschema.Schema {
+	return &jsonschema.Schema{
+		Type:     []jsonschema.Type{jsonschema.ObjectType},
+		Required: []string{cloud.EndpointKey, cloud.AuthTypesKey, cloud.RegionsKey},
+		Order:    []string{cloud.EndpointKey, cloud.AuthTypesKey, cloud.RegionsKey},
+		Properties: map[string]*jsonschema.Schema{
+			cloud.EndpointKey: {
+				Singular: "the API endpoint url for the cloud",
+				Type:     []jsonschema.Type{jsonschema.StringType},
+				Format:   jsonschema.FormatURI,
+			},
+			cloud.AuthTypesKey: &jsonschema.Schema{
+				// don't need a prompt, since there's only one choice.
+				Type: []jsonschema.Type{jsonschema.ArrayType},
+				Enum: []interface{}{[]string{string(cloud.UserPassAuthType)}},
+			},
+			cloud.RegionsKey: {
+				Type:     []jsonschema.Type{jsonschema.ObjectType},
+				Singular: "region",
+				Plural:   "regions",
+				AdditionalProperties: &jsonschema.Schema{
+					Type:          []jsonschema.Type{jsonschema.ObjectType},
+					MaxProperties: jsonschema.Int(0),
+				},
+			},
+		},
+	}
+}
+
+// Ping tests the connection to the oracle cloud to verify the endoint is valid.
+func (e environProvider) Ping(endpoint string) error {
+	return nil
+}
+
+// PrepareConfig prepares the configuration for the new model, based on
+// the provided arguments. PrepareConfig is expected to produce a
+// deterministic output
+func (e environProvider) PrepareConfig(config environs.PrepareConfigParams) (*config.Config, error) {
+	if err := e.checkSpec(config.Cloud); err != nil {
+		return nil, errors.Annotatef(err, "validating cloud spec")
+	}
+	return config.Config, nil
+}
+
+// checkSpec will try and see if the config cloud that is generated is on point with the cloudspec.
+func (e environProvider) checkSpec(spec environs.CloudSpec) error {
```

Comment:
We have reasonably well established names and patterns, like "validateCloudSpec". Please stick with the existing naming unless there's a good reason to change. Sticking with established naming and patterns makes it easier to understand one provider when you're familiar with another.